### DISCUSSION TITLE
Connecting a raspberry-pi 4B via USB to run fluidd?

### DISCUSSION THREAD

**Discussion by jeffalanmcclain** [xblax/flashforge_ad5m_klipper_mod#216] [General]
Just wondering what is required to get a raspberry-pi off-loading some of the fluidd and moonraker (mainsail) functions and timelapse functions?  I have everything working on 5M Pro with the v00.05-beta mod, and LOVE the work, however, it is a bit limited (and sometimes crashes with a timer error) if doing too much.  So, I have an r-pi with all of that KIAUH install of stuff ready to rock and use that with a minimal Klipper install on the 5M Pro.  I have no idea, however, how to connect the two.  I tried plugging into the USB port...no go, or at least I can't get any klipper function connection.  Ideas?  TIA and thanks for all the MOD work here getting all that working natively in a small memory footprint on the 5M!

**Comment by consp** [comment]
There are two ways:

1. Cut off the MCU of the AD5M and use the Pi as the controller, in which case you do not need this mod and need to connect the UARTs of the boards to the Pi (not user friendly someone did this but I'm not sure where I read it).
2. Connect the Pi to the network and run fluidd and moonraker over network and disable them on the MCU of the AD5M.

Connecting the PI via USB to the AD5M will unlikely do much.

For 2. do the following, but it would be best if you know what these commands do:

* Connect the pi to the same network as the AD5M
* Connect via a terminal to the AD5M (see docs):
* Disable the start of mainsail/fluidd
```
mv /etc/init.d/S70/fluidd /etc/init.d/K70fluidd
mv /etc/init.d/S70/mainsail /etc/init.d/K70mainsail
```
* reboot

Then connect the PI's mainsail/fluidd to the IP/port of moonraker on the AD5M.

All other options to offload more are harder to do. You might also want to disable the camera if you do not need it which can be done with:
```
mv /etc/init.d/S75camera /etc/init.d/K75camera
```

You can of course put the PI inside the case and connect it via a 24v to 5v converter to the same power source. If you do not use the ethernet you might be able to connect them together but that requires a bit more work so I would stick to wifi if you want it the easy way.

Then again with this method there is no real reason to put the PI inside since it always connects via LAN/WIFI anyways.

**Reply by jeffalanmcclain** [reply]
Thank you!  This is perfect and makes sense.  I really appreciate it....and just for completeness, I think the **mv** should be S70fluidd (no directory separator) as I look at it if anyone else tries this:

`mv /etc/init.d/S70mainsail /etc/init.d/K70mainsail`
`mv /etc/init.d/S70fluidd /etc/init.d/K70fluidd`

BR,
Jeff

**Reply by jeffalanmcclain** [reply]
Hum...ok, and just so I understand a bit better, I need to have moonraker and fluidd/mainsail installed on the r-pi, and I need to configure moonraker to connect to the remote IP of the printer?  Or how do I configure fluidd to connect to the FF printer's IP rather than the local r-pi services?  I really couldn't find a way to configure fluidd to connect to a remote IP address, just looking through the docs, and KIUAH seems to make it sound like you would use moonraker to connect to multiple (or remote) printers?

TIA!

**Reply by jeffalanmcclain** [reply]
Just a follow up...as I suspected, I should be able to just install moonraker/fluidd on the r-pi and have it [moonraker on pi] connect to the klipper running on the printer IP.  I *think* that *also* means I should NOT need to be running a moonraker instance on the printer, so I can probably terminate that now too, and then I can move to a dedicated r-pi camera for the timelapse and pull that off the printer.  All that should SIGNIFICANTLY reduce the overhead needed on the printer.  I think I will also be able to get HDMI klipper-screen working off r-pi...infact, I already have!  remotely!  It is *all* working.  Fluidd and Mainsail on r-pi (through moonraker on r-pi).  

BTW, the KIAUH script is pure $h!t for trying to do custom headless r-pi setup like this.  You don't *NEED* klipper on the r-pi for this type of network managed connection, but it pukes if you don't have it on the r-pi for the moonraker setup.  And, god forbid if you already have something (like ADSB serving) running on port 80, it just blindly fails / stomps the mainsail install.  So, to get it all working, you just go through, install Klipper, moonraker, mainsail, fluidd...then go back and change the moonraker.conf to your printer IP, change the port in the nginx setup for mainsail (it SURE WOULD BE NICE if they gave you the option to over-ride the mainsail port and moonraker IP in the script).  Then you're free to remove klipper from your r-pi.  All working great.  I *thought* I would be able to move the moonraker service on printer too, but that seems to be what everything on the r-pi is using to connect to the printer, so now, I'm curious if I could remove moonraker on the r-pi.  hum.

**Reply by consp** [reply]
>  I should NOT need to be running a moonraker instance on the printer, so I can probably terminate that now too

Possibly but I'm not the most experienced with doing that, should be easy enough to test I guess.

> KIAUH ... etc

Never used it, I'm pretty sure it's meant for single-device installs and pretty much idiot proofing it which likely means customization is not considered.

> You don't NEED klipper on the r-pi for this type of network managed connections

Correct afaik.

> so now, I'm curious if I could remove moonraker on the r-pi. hum.

I think you only need one. Either on the printer or on the r-pi, but I'm outside my zone of expertise here.

As long as it's still setup and klipper is running I see no point in just tying it out and see how it goes :)

**Reply by 0parenthesis** [reply]
Any new findings with this? and possibly a detailed tutorial for an idiot like me?

**Reply by FJay81** [reply]
yes i would like some kind of tutorial, i dont want to do something wrong and not know what

**Reply by AJolly** [reply]
https://github.com/g992/flashforge-ad5m-5mpro-research/blob/main/upgrades/klipper_on_ext_cpu.md  
This has the instructions.  
But I would love a simpler way. @jeffalanmcclain were you able to get it working over the network?

**Reply by jeffalanmcclain** [reply]
Yes!  You only need Fluidd or Mainsail installed on RPi, and you connect to your Moonraker session from the printer IP.  But the KIAUH script screws up config royally for this, so I recommend you install Klipper, Mainsail and Fluidd all and then go back in and adjust the config settings files on your rpi.  You can then also terminate those on your FlashForge 5M as CONSP instructed above.  No hardware changes needed from me.  I'm not sure it lowers the system load that much only having FLUIDD/MAINSAIL on RPi, but the video creation then I think is all done on the rpi for timelapse at least.  I still get an OCCASIONAL "MCU timer error" crash on AD5M, but I think that has something to do with the old(ish) branch of Klipper that is running.  I was ONLY able to get camera working in Fluidd via HTTP protocol (which mainsail didn't have) and put in the IP with the /action=stream part, but it is working.

**Reply by jeffalanmcclain** [reply]
FWIW, this has been working for quite some time now, and I'm happy with it.  However, it has NOT stopped MCU Shutdown: timer too close errors from occuring on the limited resources printer.  Timelapses are still run on the printer as well as USB web cam streaming.  I'm quite frustrated with the limitations of the hardware on this printer's controller (overall, the printer is quite nice and works well, except for such limited hardware).  If I had a spare old Voron controller, I might try to replace the guts of this printer with completely new controller (I may *still* order a nicer new one just to learn).  I have now put a second chamber fan in the system on the right to help PETG printing and would LOVE to have a chamber thermister...at some point, I just need to build myself a Voron Trident.  :/

**Reply by NicoletoAerials** [reply]
@jeffalanmcclain Hi! I recently got my AD5M Pro, and I'm really enjoying it, especially the quick-change nozzle system. However, like you, I'm not a fan of the hardware limitations. I have a BTT Manta M8P board lying around and am considering replacing the AD5M board with it. My main concern is how to properly connect the e-board to control the extruder and nozzle heater. Do you have any advice or suggestions?

**Reply by jeffalanmcclain** [reply]
I don't.  Sorry.  Unfortunately, there is a lot of custom things you'll probably have to do there.  I just finished building a 350mm Siboor Trident AWD, and the freedom and flexibility is quite a nice contrast to how locked down even this KlipperMOD was.  Best of luck!

**Reply by Toniob** [reply]
I didn’t buy a AD5M yet. But I’m already looking for ways to improve it. Did someone tried to export the mcu devices (/dev/tty*) on the network with socat ? If those devices are available on the network, it may be possible to run Moonraker and Klipper on another host. What do you think ?

**Reply by maoravi456** [reply]
@Toniob  yes I’ve tried that..
The 5M firmware based on builtroot system, which is closed system.. if you want to install any package you need to recompile the entire firmware…
If you know how to do this - please show me how😅 it would be very helpful (my main goal is to connect Palette 2 via hub to klipper)

**Reply by Toniob** [reply]
@maoravi456 That’s a bummer. From what I understand, python is available on the printer. Maybe using a socat alternative in python ?

**Reply by Toniob** [reply]
I’ve pushed a PR to make Klipper listen to a TCP Socket : https://github.com/Klipper3d/klipper/pull/6826 . That way, we could run Moonraker, fluid and Mailsail somewhere else. We can free ressources that way (cpu and ram).
